### PR TITLE
Fix right outer join with limit error plan

### PR DESCRIPTION
**Pull request by Seaven**
#741

Fix bug:
* JoinAssociativityRule: can't work if child join contains limit, will cause result error
* Join subquery which with limit, the shuffle way error, will cause result error(right join right child contains limit)

### CODE REVIEW

**Review by kangkaisen — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/ChildPropertyDeriver.java [hunk 1]
```diff
@@ -110,24 +111,8 @@ public Void visitPhysicalHashJoin(PhysicalHashJoinOperator node, ExpressionConte
...
-                outputInputProps.add(new Pair<>(PhysicalPropertySet.EMPTY,
-                        Lists.newArrayList(new PhysicalPropertySet(), rightBroadcastProperty)));
-            }
-            // If child has limit, only do broadcast join
-            return visitJoinRequirements(node, context, false);
-        } else {
-            outputInputProps.add(new Pair<>(PhysicalPropertySet.EMPTY,
-                    Lists.newArrayList(new PhysicalPropertySet(), rightBroadcastProperty)));
-        }
+        outputInputProps.add(new Pair<>(PhysicalPropertySet.EMPTY,
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/ChildPropertyDeriver.java:114]: right outer join/full outer join could generate any/broadcast inputProperties?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/JoinForceLimitRule.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+        long nodeLimit = join.hasLimit() ? join.getLimit() : Long.MAX_VALUE;
+        long leftLimit = left.hasLimit() ? left.getLimit() : Long.MAX_VALUE;
+        long rightLimit = right.hasLimit() ? right.getLimit() : Long.MAX_VALUE;
+
+        OptExpression newLeft = input.getInputs().get(0);
+        OptExpression newRight = input.getInputs().get(1);
+
+        if (nodeLimit > leftLimit) {
+            newLeft = OptExpression.create(new LogicalLimitOperator(left.getLimit()), newLeft);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/JoinForceLimitRule.java:57]: offset lost

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 3]
```diff
@@ -1662,5 +1662,10 @@ public PlanFragment visitPhysicalTableFunction(OptExpression optExpression, Exec
             inputFragment.setPlanRoot(tableFunctionNode);
             return inputFragment;
         }
+
+        @Override
+        public PlanFragment visitPhysicalLimit(OptExpression optExpression, ExecPlan context) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1667]: Would better add some comments here

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/JoinForceLimitRule.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    public boolean check(OptExpression input, OptimizerContext context) {
+        Operator join = input.getOp();
+        Operator left = input.getInputs().get(0).getOp();
+        Operator right = input.getInputs().get(1).getOp();
+
+        long nodeLimit = join.hasLimit() ? join.getLimit() : Long.MAX_VALUE;
+        long leftLimit = left.hasLimit() ? left.getLimit() : Long.MAX_VALUE;
+        long rightLimit = right.hasLimit() ? right.getLimit() : Long.MAX_VALUE;
+
+        return nodeLimit > leftLimit || nodeLimit > rightLimit;
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/JoinForceLimitRule.java]: I think don't need.
1. Offset limit will be merge with sort to TopN
2. Can't repeat choose offset if child is TopN

File: fe/fe-core/src/test/resources/sql/subquery/exists-subquery.sql [hunk 5]
```diff
@@ -4,7 +4,8 @@ select t0.v1 from t0 where exists (select t3.v2 from t3)
 CROSS JOIN (join-predicate [null] post-join-predicate [null])
     SCAN (columns[1: v1] predicate[null])
     EXCHANGE BROADCAST
-        SCAN (columns[4: v1] predicate[null]) Limit 1
+        EXCHANGE GATHER
```
Comment (Youngwb) [inline fe/fe-core/src/test/resources/sql/subquery/exists-subquery.sql:7]: why have a gather here

**Review by Youngwb — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**